src/Surfnet/StepupSelfService/SelfServiceBundle/Security/Authentication/Handler/ExplicitSessionTimeoutHandler.php

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupSelfService\SelfServiceBundle\Security\Authentication\Handler;
+
+use Psr\Log\LoggerInterface;
+use Surfnet\StepupSelfService\SelfServiceBundle\Security\Authentication\AuthenticatedSessionStateHandler;
+use Surfnet\StepupSelfService\SelfServiceBundle\Security\Authentication\Session\SessionLifetimeGuard;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Http\Logout\CookieClearingLogoutHandler;
+use Symfony\Component\Security\Http\Logout\SessionLogoutHandler;
+
+class ExplicitSessionTimeoutHandler implements AuthenticationHandler
+{
+    /**
+     * @var AuthenticationHandler|null
+     */
+    private $nextHandler;
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var SessionLifetimeGuard
+     */
+    private $sessionLifetimeGuard;
+
+    /**
+     * @var AuthenticatedSessionStateHandler
+     */
+    private $authenticatedSession;
+
+    /**
+     * @var SessionLogoutHandler
+     */
+    private $sessionLogoutHandler;
+
+    /**
+     * @var CookieClearingLogoutHandler
+     */
+    private $cookieClearingLogoutHandler;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(
+        TokenStorageInterface $tokenStorageInterface,
+        AuthenticatedSessionStateHandler $authenticatedSessionStateHandler,
+        SessionLifetimeGuard $sessionLifetimeGuard,
+        SessionLogoutHandler $sessionLogoutHandler,
+        CookieClearingLogoutHandler $cookieClearingLogoutHandler,
+        RouterInterface $router,
+        LoggerInterface $logger
+    ) {
+        $this->tokenStorage                = $tokenStorageInterface;
+        $this->authenticatedSession        = $authenticatedSessionStateHandler;
+        $this->sessionLifetimeGuard        = $sessionLifetimeGuard;
+        $this->sessionLogoutHandler        = $sessionLogoutHandler;
+        $this->cookieClearingLogoutHandler = $cookieClearingLogoutHandler;
+        $this->router                      = $router;
+        $this->logger                      = $logger;
+    }
+
+    public function process(GetResponseEvent $event)
+    {
+        if ($this->tokenStorage->getToken() !== null
+            && !$this->sessionLifetimeGuard->sessionLifetimeWithinLimits($this->authenticatedSession)
+        ) {
+            $invalidatedBy = [];
+            if ($this->sessionLifetimeGuard->sessionLifetimeWithinAbsoluteLimit($this->authenticatedSession)) {
```

Comment:
The session will be logged as invalidated by the absolute limit when the lifetime is within the limit instead of outside the limit, I think this check (and the one below) should be inverted.